File: problems-originals/0401-0500/0448_find-all-numbers-disappeared-in-an-array/solution_hash_set.go
// The direct reading: record every value in a hash set, then walk the
// candidate range 1..n and keep the values the set does not hold. The set
// carries no order of its own; walking the candidates in increasing order
// is what makes the pinned ascending output free.
func findDisappearedNumbers(nums []int) []int {
	// make (never a nil slice) so an empty answer marshals as [] not null;
	// len(nums) is the output's upper bound: at most n-1 values disappear.
	disappeared := make([]int, 0, len(nums))
	seen := make(map[int]bool, len(nums))
	for _, value := range nums {
		seen[value] = true
	}
	for value := 1; value <= len(nums); value++ {
		if !seen[value] {
			disappeared = append(disappeared, value)
		}
	}
	return disappeared
}

File: problems-adapt/0401-0500/0448_missing-range-values/solution_hash_set.cpp
class Solution {
  public:
    vector<int> missingValues(vector<int> &nums) {
        // The direct reading: record every value in a hash set, then walk the
        // candidate range 1..n and keep the values the set does not hold.
        unordered_set<int> seen(nums.begin(), nums.end());
        // The set carries no order of its own; walking the candidates in
        // increasing order is what makes the pinned ascending output free.
        vector<int> disappeared;
        for (int value = 1; value <= (int)nums.size(); ++value) {
            if (!seen.count(value)) {
                disappeared.push_back(value);
            }
        }
        return disappeared;
    }
};
